support/cas-server-support-oauth/src/main/java/org/apereo/cas/support/oauth/profile/ClientIdAwareProfileManager.java

```diff
@@ -0,0 +1,55 @@
+package org.apereo.cas.support.oauth.profile;
+
+import com.github.scribejava.core.model.OAuthConstants;
+import org.pac4j.core.context.WebContext;
+import org.pac4j.core.context.session.SessionStore;
+import org.pac4j.core.profile.CommonProfile;
+import org.pac4j.core.profile.ProfileManager;
+
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * This is a profile manager used during OAuth authentication.
+ *
+ * It saves returns a profile based on client_id from the request.
+ *
+ * @author Kirill Gagarski
+ * @param <U> profile type
+ * @since 5.3.8
+ */
+public class ClientIdAwareProfileManager<U extends CommonProfile> extends ProfileManager<U> {
+
+    private static final String SESSION_CLIENT_ID = "oauthClientId";
+    private static final String REQUEST_CLIENT_ID = OAuthConstants.CLIENT_ID;
+
+    public ClientIdAwareProfileManager(final WebContext context, final SessionStore sessionStore) {
+        super(context, sessionStore);
+    }
+
+    public ClientIdAwareProfileManager(final WebContext context) {
+        super(context);
+    }
+
+    @Override
+    protected LinkedHashMap<String, U> retrieveAll(final boolean readFromSession) {
+        return super.retrieveAll(readFromSession).entrySet().stream().filter(
+            it -> it.getValue()
+                .getAuthenticationAttribute(SESSION_CLIENT_ID)
+                .equals(context.getRequestParameter(REQUEST_CLIENT_ID)))
+            .collect(Collectors.toMap(
+                Map.Entry::getKey,
+                Map.Entry::getValue,
+                (v1, v2) -> {
+                    throw new IllegalStateException("Duplicate key");
+                },
+                LinkedHashMap::new));
+    }
+
+    @Override
+    public void save(final boolean saveInSession, final U profile, final boolean multiProfile) {
+        profile.addAuthenticationAttribute(SESSION_CLIENT_ID, context.getRequestParameter(REQUEST_CLIENT_ID));
+        super.save(saveInSession, profile, multiProfile);
```

Comment:
Maybe we should enforce `multiProfile` here.

Now If we subsequently authenticate into A, then into B, then into A, then into B, we go through full redirect chain (authorize -> callbackAuthorize -> login -> authorize) each time. Multiprofile might prevent it but I am not sure that this is a good idea.

Reply:
It will work in both cases.
If you enable `multiProfile`, all profiles (each attached to a `client_id`) will be saved into the web session and with your mechanism, you'll pick up the right one.
If you disable `multiProfile`,  only one profile (attached to a `client_id`) will be saved into the web session and erase the previous one. With your mechanism, you'll return no profile if it's not the right `client_id` starting a new login process.
Option 1 should provide better performance, but I fear side effects if we keep many profiles in session, especially at logout. I'd prefer to disable `multiProfile`.